--- a/tools/setup_helpers/extension.py	2025-01-29 02:11:55.000000000 +0100
+++ b/tools/setup_helpers/extension.py	2025-04-02 03:31:43.995743094 +0200
@@ -140,6 +140,7 @@
             f"-DUSE_CUDA:BOOL={'ON' if _USE_CUDA else 'OFF'}",
             f"-DUSE_OPENMP:BOOL={'ON' if _USE_OPENMP else 'OFF'}",
             f"-DUSE_FFMPEG:BOOL={'ON' if _USE_FFMPEG else 'OFF'}",
+            "-DCMAKE_POLICY_VERSION_MINIMUM=3.5",
         ]
         build_args = ["--target", "install"]
         # Pass CUDA architecture to cmake
