diff --git llama_cpp/llama_cpp.py llama_cpp/llama_cpp.py
index 21f85c81..cc35670f 100644
--- llama_cpp/llama_cpp.py
+++ llama_cpp/llama_cpp.py
@@ -35,9 +35,7 @@ if TYPE_CHECKING:
 _lib_base_name = "llama"
 _override_base_path = os.environ.get("LLAMA_CPP_LIB_PATH")
 _base_path = (
-    pathlib.Path(os.path.abspath(os.path.dirname(__file__))) / "lib"
-    if _override_base_path is None
-    else pathlib.Path(_override_base_path)
+    pathlib.Path("/usr/lib") if _override_base_path is None else pathlib.Path(_override_base_path)
 )
 # Load the library
 _lib = load_shared_library(_lib_base_name, _base_path)
