--- llama_cpp/llama_cpp.py
+++ llama_cpp/llama_cpp.py
@@ -33,7 +33,7 @@
 # Specify the base name of the shared library to load
 _lib_base_name = "llama"
 _override_base_path = os.environ.get("LLAMA_CPP_LIB_PATH")
-_base_path = pathlib.Path(os.path.abspath(os.path.dirname(__file__))) / "lib" if _override_base_path is None else pathlib.Path(_override_base_path)
+_base_path = pathlib.Path("/usr/lib") if _override_base_path is None else pathlib.Path(_override_base_path)
 # Load the library
 _lib = load_shared_library(_lib_base_name, _base_path)
