diff --git a/CMakeLists.txt b/CMakeLists.txt
index 515b91be1..9962c1c16 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -212,6 +212,7 @@ install(TARGETS whisper LIBRARY PUBLIC_HEADER)
 
 target_compile_definitions(whisper PRIVATE
     WHISPER_VERSION="${WHISPER_VERSION}"
+    WHISPER_BUILD_COMMIT="${WHISPER_BUILD_COMMIT}"
 )
 
 set_target_properties(parakeet PROPERTIES PUBLIC_HEADER ${CMAKE_CURRENT_SOURCE_DIR}/include/parakeet.h)
diff --git a/src/whisper.cpp b/src/whisper.cpp
index 89146e2e4..23b434d76 100644
--- a/src/whisper.cpp
+++ b/src/whisper.cpp
@@ -9170,7 +9170,11 @@ void whisper_log_set(ggml_log_callback log_callback, void * user_data) {
 }
 
 const char * whisper_version(void) {
+#ifdef WHISPER_BUILD_COMMIT
+    return WHISPER_VERSION " (" WHISPER_BUILD_COMMIT ")";
+#else
     return WHISPER_VERSION;
+#endif
 }
 
 GGML_ATTRIBUTE_FORMAT(2, 3)
